### PR TITLE
feat: gRPC tracking

### LINKED ISSUE #2991: chore: tracking for gRPC
### Feature request

Prometheus interceptor has some difference in metrics implementation, hence we should revisit analytics once more.

### Motivation

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

## What does this PR address?

Fixes #2991 
Depends on #2984 

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: bentoml/_internal/utils/__init__.py [hunk 1]
```diff
@@ -350,13 +354,16 @@ class cached_contextmanager:
...
+
+    Example Usage: (To reuse the container based on the same image)
+
+    .. code-block:: python
+
+        @cached_contextmanager("{docker_image.id}")
+        def start_docker_container_from_image(docker_image, timeout=60):
+            container = ...
+            yield container
+            container.stop()
```
Comment (aarnphm) [inline bentoml/_internal/utils/__init__.py:366]: update documentation how to use this decorator.

File: bentoml/_internal/utils/analytics/usage_stats.py [hunk 2]
```diff
@@ -62,7 +66,12 @@ def wrapper(*args: P.args, **kwargs: P.kwargs) -> t.Any:
             return func(*args, **kwargs)
         except Exception as err:  # pylint: disable=broad-except
             if _usage_event_debugging():
-                logger.info(f"Tracking Error: {err}")
+                if get_debug_mode():
+                    logger.error(
+                        f"Tracking Error: {err}", stack_info=True, stacklevel=3
+                    )
+                else:
+                    logger.info(f"Tracking Error: {err}")
```
Comment (aarnphm) [inline bentoml/_internal/utils/analytics/usage_stats.py:74]: Showing stack when `BENTOML_DEBUG=True`

File: bentoml/_internal/server/metrics/prometheus.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def prometheus_client(self):
 
         # step 2:
         import prometheus_client
+        import prometheus_client.parser
```
Comment (aarnphm) [inline bentoml/_internal/server/metrics/prometheus.py:59]: This changes is to make sure we don't import prometheus_client.parser again.

File: bentoml/_internal/utils/analytics/usage_stats.py [hunk 4]
```diff
@@ -204,27 +250,33 @@ def track_serve(
...
     stop_event = threading.Event()
 
     @silent
     def loop() -> t.NoReturn:  # type: ignore
         last_tracked_timestamp: datetime = serve_info.serve_started_timestamp
         while not stop_event.wait(tracking_interval):  # pragma: no cover
+            metrics_type = "not_available"
+            metrics, use_legacy_metrics = get_metrics_report(metrics_client, grpc=grpc)
+            if use_legacy_metrics is not None:
+                metrics_type = "legacy" if use_legacy_metrics else "current"
```
Comment (aarnphm) [inline bentoml/_internal/utils/analytics/usage_stats.py:269]: I'm not sure if we want to call the metrics type `legacy` and `current`. Maybe we can call it v1 and v2? @ssheng

File: bentoml/_internal/utils/analytics/schemas.py [hunk 5]
```diff
@@ -199,9 +200,15 @@ class ServeInitEvent(EventMeta):
 class ServeUpdateEvent(EventMeta):
     serve_id: str
     production: bool
-    component: str
+    grpc: bool
```
Comment (ssheng) [inline bentoml/_internal/utils/analytics/schemas.py:203]: Can we make this field a string to be expandable in the future? `http`, `grpc`, `job`, etc.

Comment (aarnphm) [reply bentoml/_internal/utils/analytics/schemas.py:203]: lets do that

**Review by ssheng — APPROVED** [review]

**Pull request merged**